packages/datetime/src/dateRangeInput.tsx

```diff
@@ -326,19 +366,20 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
             } else {
                 this.setState({ [keys.inputString]: "", [keys.selectedValue]: moment(null) });
             }
-            Utils.safeInvoke(this.props.onChange, [null, null] as DateRange);
+            Utils.safeInvoke(this.props.onChange, this.getDateRangeForCallback(moment(null), boundary));
         } else if (this.isMomentValidAndInRange(maybeNextValue)) {
+            // note that error cases that depend on both fields (e.g.
+            // overlapping dates) should fall through into this block so that
+            // the UI can update immediately, possibly with an error message on
+            // the other field.
             if (isValueControlled) {
                 this.setState({ [keys.inputString]: inputString });
             } else {
                 this.setState({ [keys.inputString]: inputString, [keys.selectedValue]: maybeNextValue });
             }
```

Comment:
this `if (controlled) { setState(A); } else { setState(A + B); }` pattern happens a lot in this class. seems like a good opportunity for a helper function 🤔  though perhaps a separate PR.